File: test_doubles/csharp/TestDoubles/TestDoubles/Service/PurchaseService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using TestDoubles.Domain;
using TestDoubles.Repository;

namespace TestDoubles
{
    public class PurchaseService
    {
        private PurchaseRepository purchaseRepository;

        public Invoice MakePurchase(Customer customer, 
            Product product, Discount discount)
        {
            var purchase 
                = purchaseRepository.CreatePurchase(customer);
            purchaseRepository.AddProduct(purchase, product);
            var invoice 
                = purchaseRepository.CreateInvoice(purchase);

            if (discount != null)
            {
                invoice.ApplyDiscount(discount);
            }
            return invoice;
        }
    }
}

File: test_doubles/java/TestDoubles/src/main/java/service/PurchaseService.java
package service;

import domain.*;
import repository.PurchaseRepository;

public class PurchaseService {
    private PurchaseRepository purchaseRepository;

    public Invoice MakePurchase(Customer customer, Product product, Discount discount) {
        Purchase purchase = purchaseRepository.createPurchase(customer);
        purchaseRepository.addProduct(purchase, product);
        Invoice invoice = purchaseRepository.createInvoice(purchase);

        if (discount != null) {
            invoice.applyDiscount(discount);
        }
        return invoice;
    }
}

